kani-compiler/src/codegen_cprover_gotoc/codegen/foreign_function.rs

```diff
@@ -0,0 +1,155 @@
+            self.ensure(fn_name, |gcx, _| {
+                let typ = gcx.codegen_ffi_type(instance);
+                Symbol::function(
+                    fn_name,
+                    typ,
+                    None,
+                    gcx.readable_instance_name(instance),
+                    Location::none(),
+                )
+                .with_is_extern(true)
+            })
+        } else {
+            let shim_name = format!("{fn_name}_ffi_shim");
+            trace!(?shim_name, "codegen_foreign_function");
+            self.ensure(&shim_name, |gcx, _| {
+                // Generate a shim with an unsupported C-FFI error message.
+                let typ = gcx.codegen_ffi_type(instance);
+                Symbol::function(
+                    &shim_name,
+                    typ,
+                    Some(gcx.codegen_ffi_shim(shim_name.as_str().into(), instance)),
+                    gcx.readable_instance_name(instance),
+                    Location::none(),
+                )
+            })
+        }
+    }
+
+    /// Checks whether C-FFI has been enabled or not.
+    /// When enabled, we blindly encode the function type as is.
+    fn is_cffi_enabled(&self) -> bool {
+        self.queries.get_unstable_features().contains(&"c-ffi".to_string())
+    }
+
+    /// Generate code for a foreign function shim.
+    fn codegen_ffi_shim(&mut self, shim_name: InternedString, instance: Instance<'tcx>) -> Stmt {
+        debug!(?shim_name, ?instance, sym=?self.symbol_table.lookup(shim_name), "generate_foreign_shim");
+
+        let loc = self.codegen_span(&self.tcx.def_span(instance.def_id()));
+        let unsupported_check = self.codegen_ffi_unsupported(instance, loc);
+        Stmt::block(vec![unsupported_check], loc)
+    }
+
+    /// Generate type for the given foreign instance.
+    fn codegen_ffi_type(&mut self, instance: Instance<'tcx>) -> Type {
+        let fn_name = self.symbol_name(instance);
+        let fn_abi = kani_middle::fn_abi(self.tcx, instance);
+        let loc = self.codegen_span(&self.tcx.def_span(instance.def_id()));
+        let params = fn_abi
+            .args
+            .iter()
+            .enumerate()
+            .filter_map(|(idx, arg)| {
+                (!arg.is_ignore()).then(|| {
+                    let arg_name = format!("{fn_name}::param_{idx}");
+                    let base_name = format!("param_{idx}");
+                    let arg_type = self.codegen_ty(arg.layout.ty);
+                    let sym = Symbol::variable(&arg_name, &base_name, arg_type.clone(), loc)
+                        .with_is_parameter(true);
+                    self.symbol_table.insert(sym);
+                    arg_type.as_parameter(Some(arg_name.into()), Some(base_name.into()))
+                })
+            })
+            .collect();
+        let ret_type = self.codegen_ty(fn_abi.ret.layout.ty);
+
+        if fn_abi.c_variadic {
+            Type::variadic_code(params, ret_type)
```

Comment:
Do we have a test case for variadic code?

Reply:
Yes, we have a fix-me test today. However, it wasn't really checking what it was meant to. I'll create a separate PR that fixes that test since this PR is already big enough.

Reply:
btw, I fixed the test here, but what I think we should do later is turn this test into an expected test and make sure we are emitting the right error.